react/features/base/app/components/BaseApp.js

```diff
@@ -0,0 +1,255 @@
+ *
+ * @abstract
+ */
+export default class BaseApp extends Component<*, State> {
+    _init: Promise<*>;
+
+    /**
+     * Initializes a new {@code BaseApp} instance.
+     *
+     * @param {Object} props - The read-only React {@code Component} props with
+     * which the new instance is to be initialized.
+     */
+    constructor(props: Object) {
+        super(props);
+
+        this.state = {
+            initialized: false,
+            route: {},
+
+            // $FlowFixMe
+            store: undefined
+        };
+
+        /**
+         * Make the mobile {@code BaseApp} wait until the
+         * {@code AsyncStorage} implementation of {@code Storage} initializes
+         * fully.
+         *
+         * @private
+         * @see {@link #_initStorage}
+         * @type {Promise}
+         */
+        this._init
+            = this._initStorage()
+                .catch(() => { /* AbstractApp should always initialize! */ })
+                .then(() =>
+                    this.setState({
+                        store: this._createStore()
+                    }));
+    }
+
+    /**
+     * Initialize the application.
+     *
+     * @inheritdoc
+     */
+    componentWillMount() {
+        this._init.then(() => {
+            const { dispatch } = this.state.store;
+
+            dispatch(appWillMount(this));
+
+            // We set the initialized state here and not in the constructor to
+            // make sure that {@code componentWillMount} gets invoked before
+            // the app tries to render the actual app content.
+            this.setState({ initialized: true });
+        });
+    }
+
+    /**
+     * De-initialize the application.
+     *
+     * @inheritdoc
+     */
+    componentWillUnmount() {
+        const { dispatch } = this.state.store;
+
+        dispatch(appWillUnmount(this));
+    }
+
+    /**
+     * Delays this {@code BaseApp}'s startup until the {@code Storage}
+     * implementation of {@code localStorage} initializes. While the
+     * initialization is instantaneous on Web (with Web Storage API), it is
+     * asynchronous on mobile/react-native.
+     *
+     * @private
+     * @returns {Promise}
+     */
+    _initStorage(): Promise<*> {
+        const { _initializing } = window.localStorage;
+
+        return _initializing || Promise.resolve();
+    }
+
+    /**
+     * Implements React's {@link Component#render()}.
+     *
+     * @inheritdoc
+     * @returns {ReactElement}
+     */
+    render() {
+        const { initialized, route, store } = this.state;
+        const { component } = route;
+
+        if (initialized && component) {
+            return (
+                <I18nextProvider i18n = { i18next }>
+                    <Provider store = { store }>
+                        <Fragment>
+                            { this._createMainElement(component) }
+                            <SoundCollection />
+                            { this._createExtraElement() }
```

Comment:
I did this so `App` could attach the `OverlayContainer` (and potentially more things) here, since that's not a base feature.